Tasks/Common/packaging-common/nuget/NuGetConfigHelper2.ts

```diff
@@ -158,6 +161,19 @@ export class NuGetConfigHelper2 {
         return packageSources.map((source) => this.convertToIPackageSource(source));
     }
 
+    // TODO: Remove this once NuGet issue https://github.com/NuGet/Home/issues/7855 is fixed.
+    public backupExistingNuGetFiles(): void {
+        this.nuGetFiles = fs.readdirSync('.').filter((file) => file.toLowerCase() === nugetFileName);
```

Comment:
You're making a backup of the existing file in the root directory, but if I have a composed nuget config made of 2 nuget configs, one in the root and one elsewhere (nugetConfigPath), this will effectively remove anything defined in the root nuget config for this restore.

Reply:
I think you're on the right track, and I'm willing to accept this as a temporary fix if you can come up with a way to resolve the case - possibly just checking if nugetFileName == nugetConfigPath, and only making the backup copy if they are the same (and writing an info message if nugetFileName exists but != nugetConfigPath instead of copying the temp file)

Reply:
Thanks for the comment @jotaylo!
How are you creating the composed nuget config? I'm having issues creating the scenario you're describing:

In my source root, I have `nuget.config` and `composed_nuget.config` files. I'm using [XML External Entities to do the composition](https://www.w3.org/TR/REC-xml/#sec-external-ent).

nuget.config:
```xml
  <packageSources>
    <add key="Production" value="https://xxx.pkgs.visualstudio.com/_packaging/Production/nuget/v3/index.json" />
    <add key="PreRelease" value="https://xxx.pkgs.visualstudio.com/_packaging/PreRelease/nuget/v3/index.json" />
    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
  </packageSources>
```

composed_nuget.config:
```xml
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE doc [
<!ENTITY rootnuget SYSTEM "NuGet.config">
]>
<configuration>
  &rootnuget;
</configuration>
```

The task currently copies the referenced config file (`composed_nuget.config` to `%Agent_BuildDirectory%/Nuget/tempNuGet_%build_buildId%.config`) without doing any resolving:
```xml
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE doc [
<!ENTITY rootnuget SYSTEM "NuGet.config"> <!-- also tried absolute path to NuGet.config here. -->
]>
<configuration>
  &rootnuget;
</configuration>
```
Instead the task flags the file as being an invalid NuGet.config file:
```
...
##vso[task.debug]Setting auth in the temp nuget.config
Saving NuGet.config to a temporary config file.
##vso[task.debug]testing directory 'C:\tmp\build\Nuget'
##vso[task.debug]testing directory 'C:\tmp\build'
##vso[task.debug]mkdir 'C:\tmp\build\Nuget'
##vso[task.debug]Getting sources from NuGet.config in this location: C:\tmp\build\Nuget\tempNuGet_undefined.config
##vso[task.issue type=error;]Error: The NuGet.config at C:\tmp\build\Nuget\tempNuGet_undefined.config is invalid.
##vso[task.debug]task result: Failed
##vso[task.issue type=error;]Packages failed to restore
##vso[task.complete result=Failed;]Packages failed to restore
```

Files composed with XML External Entities don't appear to be supported.

Is there a composition method that works? If so can I have instructions so I can consider this scenario?

Reply:
Did you mean a composed_nuget.config that inherited from the working directory's nuget.config?
According to the [Configuring the behaviour of NuGet documentation](https://docs.microsoft.com/en-us/nuget/consume-packages/configuring-nuget-behavior):

> In a solution folder [a.k.a. working directory, or root source folder], settings apply apply to all projects in subfolders. Note that if a config file is placed in a project folder [a.k.a. sub folder of the source], it has no effect on that project.

It's my understanding that when a `nugetConfigPath` is specified, this is the solution-level file described above and would ignore any other nuget files in the solution (including a source root `nuget.config`).

Reply:
See point 5 under [How settings are applied](https://docs.microsoft.com/en-us/nuget/consume-packages/configuring-nuget-behavior#how-settings-are-applied).

> For example, if a command is invoked in c:\A\B\C, NuGet looks for and loads config files in c:, then c:\A, then c:\A\B, and finally c:\A\B\C.

In our case, say nugetConfigPath is c:\A\B\C\D, and the root directory is c:\A\B\C. Since nuget.exe is invoked from the root directory, it will pick up settings from nuget.config in c:\A\B\C.

Reply:
Okay I've accounted for this now.
I added a condition to check firstly if the supplied `this.nugetConfigPath` is located in the root directory and equal to `'nuget.config'`. If it is, then we do the backup/rename/restore flow.
Otherwise, if there doesn't exist any root `'nuget.config'`, we also do the backup/rename/restore flow.
Otherwise we don't do the the backup/rename/restore flow.

This is implemented as the condition:
```
path.relative('.', this.nugetConfigPath).toLocaleLowerCase() == nugetFileName || this.rootNuGetFiles.length ==  0
```

Tested it locally and on Azure DevOps.